diff --git a/src/main.cpp b/src/main.cpp
index 20114b7..2bb949f 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -437,7 +437,7 @@ void Session::GetSupportedDecrypterURN(std::pair<std::string, std::string> &urn)
   VFSDirEntry *items(0);
   unsigned int num_items(0);
 
-  xbmc->Log(ADDON::LOG_DEBUG, "Searching for decrypters in: %s", path);
+  xbmc->Log(ADDON::LOG_DEBUG, "Searching for decrypters in: %s", path.c_str());
 
   if (!xbmc->GetDirectory(path.c_str(), "", &items, &num_items))
     return;
