From 244a4623c617afbe213a0818f5bd86faeb3986e7 Mon Sep 17 00:00:00 2001
From: Chris Mayo <aklhfex@gmail.com>
Date: Fri, 16 Dec 2016 19:22:32 +0000
Subject: [PATCH] [cmake] linux: don't strip binaries

---
 project/cmake/scripts/linux/ArchSetup.cmake | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/project/cmake/scripts/linux/ArchSetup.cmake b/project/cmake/scripts/linux/ArchSetup.cmake
index bcd70df..1e46f0f 100644
--- a/project/cmake/scripts/linux/ArchSetup.cmake
+++ b/project/cmake/scripts/linux/ArchSetup.cmake
@@ -24,11 +24,6 @@ else()
   endif()
 endif()
 
-# Make sure we strip binaries in Release build
-if(CMAKE_BUILD_TYPE STREQUAL Release AND CMAKE_COMPILER_IS_GNUCXX)
-  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -s")
-endif()
-
 find_package(CXX11 REQUIRED)
 include(LDGOLD)
 
