diff -ruN kodi-visualization-shadertoy-9999.orig/src/main.cpp kodi-visualization-shadertoy-9999/src/main.cpp
--- kodi-visualization-shadertoy-9999.orig/src/main.cpp	2016-01-26 22:09:57.477898036 +0100
+++ kodi-visualization-shadertoy-9999/src/main.cpp	2016-01-26 22:10:38.483479007 +0100
@@ -34,7 +34,7 @@
 #include <fstream>
 #include <streambuf>
 #include <ctime>
-#include "platform/util/timeutils.h"
+#include "p8-platform/util/timeutils.h"
 #include <math.h>
 #include <complex.h>
 #include <limits.h>
@@ -628,7 +628,7 @@
       needsUpload = false;
     }
 
-    float t = (float)PLATFORM::GetTimeMs() / 1000.0f;
+    float t = (float)P8PLATFORM::GetTimeMs() / 1000.0f;
     GLfloat tv[] = { t, t, t, t };
 
     glUseProgram(shader);
